src/util/ALE.sol

```diff
@@ -0,0 +1,235 @@
+            _permit.deadline,
+            _permit.v,
+            _permit.r,
+            _permit.s
+        );
+
+        if (dola.balanceOf(address(this)) < _value) revert DOLAInvalidBorrow();
+
+        // Burn the dola minted previously
+        dola.burn(_value);
+
+        // Refund any possible unspent 0x protocol fees to the sender.
+        payable(msg.sender).transfer(address(this).balance);
+    }
+
+    function deleveragePosition(
+        // Amount of DOLA to repay
+        uint256 _value,
+        // `sellTokenAddress` field from the API response.
+        address _sellTokenAddress,
+        // Collateral amount to withdraw from the escrow
+        uint256 _collateralAmount,
+        // The `allowanceTarget` field from the API response.
+        address _spender,
+        // The `to` field from the API response.
+        bytes calldata _swapCallData,
+        // Permit data
+        Permit calldata _permit,
+        // Optional helper data in case collateral needs to be transformed
+        bytes calldata _helperData
+    ) external payable nonReentrant {
+        if (markets[_sellTokenAddress].market == address(0))
+            revert MarketNotSetForCollateral(_sellTokenAddress);
+
+        IMarket market = IMarket(markets[_sellTokenAddress].market);
+        IERC20 collateral = IERC20(markets[_sellTokenAddress].collateral);
+        
+        dola.mint(address(this), _value);
+        dola.approve(address(market), _value);
+
+        market.repay(msg.sender, _value);
+
+        // withdraw amount from ZERO EX quote
+        market.withdrawOnBehalf(
+            msg.sender,
+            _collateralAmount,
+            _permit.deadline,
+            _permit.v,
+            _permit.r,
+            _permit.s
+        );
+
+        // If there's an helper contract, the collateral has to be transformed
+        if (markets[_sellTokenAddress].helper != address(0)) {
+            //Approve collateral to be transformed to sellToken
+            collateral.approve(
+                markets[_sellTokenAddress].helper,
+                _collateralAmount
+            );
+            // Collateral amount is now transformed into sellToken
+            _collateralAmount = ITransformHelper(
+                markets[_sellTokenAddress].helper
+            ).transformFromCollateral(_collateralAmount, _helperData);
+            // Approve sellToken for spender
+            IERC20(_sellTokenAddress).approve(_spender, _collateralAmount);
+        } else {
+            // If there's no helper the withdrawn collateral is the sellToken
+            if (address(collateral) == address(0)) revert CollateralNotSet();
+            // Approve collateral to be swapped
+            collateral.approve(_spender, 0); // in some cases like USDT we need to first set it to zero and the approve (TODO review this)
+            collateral.approve(_spender, _collateralAmount);
+        }
+
+        uint256 dolaBalBefore = dola.balanceOf(address(this));
```

Comment:
Is there any reason we need to check DOLA balance?

Reply:
Similar to the collateral amount above, the idea was to check that we did actually get the DOLA from the swap and not because it was directly sent to the contract, to avoid miscalculation but it can be removed, fixed in commit 311a0dc